storage/raftstorage.go

```diff
@@ -67,80 +67,88 @@ func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
 	return hs, cs, nil
 }
 
-// loadLastIndex looks in the engine to find the last log index.
+// setLastIndex persists a new last index and updates its cached
+// value in the range. If a non-nil engine is passed, it is used for the
+// write; otherwise the range's default engine is used. Commit() will
+// be called on the engine to keep the update as atomic as possible.
+func (r *Range) setLastIndex(eng engine.Engine, lastIndex uint64) error {
```

Comment:
in effect the issue here is to atomically do two atomic things. You want to commit the batch and at pretty much the next chance you get update the cached value atomically.
There are also two callsites, namely with a batch and without a batch. Take a look at the changes in `range.go`. I also think it's clumsy, hence my comment on `setLastIndex` above.